Extensions/PlayerAuthentication/playerauthenticationtools.ts

```diff
@@ -0,0 +1,833 @@
+    };
+
+    /**
+     * Retrieves the user information from the local storage, and store
+     * them in the extension variables.
+     * Returns if the fetched user is different than the current one used
+     * in the variables. (useful when the user changes)
+     */
+    const readAuthenticatedUserFromLocalStorage = (): boolean => {
+      const _gameId = gdjs.projectData.properties.projectUuid;
+
+      const authenticatedUserStorageItem = window.localStorage.getItem(
+        `${_gameId}_authenticatedUser`
+      );
+      if (!authenticatedUserStorageItem) {
+        _checkedLocalStorage = true;
+        return _userId === null;
+      }
+      const authenticatedUser = JSON.parse(authenticatedUserStorageItem);
+      const hasUserChanged = authenticatedUser.userId !== _userId;
+
+      _username = authenticatedUser.username;
+      _userId = authenticatedUser.userId;
+      _userToken = authenticatedUser.userToken;
+      _checkedLocalStorage = true;
+
+      return hasUserChanged;
+    };
+
+    /**
+     * Reads the event sent by the authentication window and
+     * display the appropriate banner.
+     */
+    const receiveMessageFromAuthenticationWindow = function (
+      runtimeScene: gdjs.RuntimeScene,
+      event: MessageEvent,
+      { checkOrigin }: { checkOrigin: boolean }
+    ) {
+      const _gameId = gdjs.projectData.properties.projectUuid;
+      const allowedOrigin = 'https://liluo.io';
+
+      // Check origin of message.
+      if (checkOrigin && event.origin !== allowedOrigin) {
+        throw new Error(`Unexpected origin: ${event.origin}`);
+      }
+      // Check that message is not malformed.
+      if (!event.data.id) {
+        throw new Error('Malformed message');
+      }
+
+      // Handle message.
+      switch (event.data.id) {
+        case 'authenticationLoaded': {
+          if (_authenticationIframe) {
+            // Hide loader.
+            if (_authenticationSubContainer && _authenticationLoader) {
+              _authenticationSubContainer.removeChild(_authenticationLoader);
+            }
+            // Show iframe
+            _authenticationIframe.style.opacity = '1';
+          }
+          break;
+        }
+        case 'authenticationResult': {
+          if (!(event.data.body && event.data.body.token)) {
+            throw new Error('Malformed message.');
+          }
+          const username = event.data.body.username;
+          if (!username) {
+            logger.warn('The authenticated player does not have a username');
+          }
+          _username = username;
+          _userId = event.data.body.userId;
+          _userToken = event.data.body.token;
+          window.localStorage.setItem(
+            `${_gameId}_authenticatedUser`,
+            JSON.stringify({
+              username: _username,
+              userId: _userId,
+              userToken: _userToken,
+            })
+          );
+
+          _justLoggedIn = true;
```

Comment:
Can this happen in the middle of the events execution?

Reply:
The callback comes from `window.addEventListener('message'`.
Does the main loop happens in another callback so they can't overstep each other or something like this?

Reply:
Unsure about the exact details, but I think so.
@4ian can you confirm?

Reply:
JS is single threaded (unless using web workers), so it means callbacks are executed one after the other, and a function who runs, like the rendering of a frame, can't have anything else "happening in between".
In particular, the frame is rendered every 16.66ms (more or less), thanks to requestAnimationFrame. During this time, no callback can run. So it means the callbacks are necessarily launched between frames:

requestAnimationFrame 1 ------> requestAnimationFrame 2 ----> callback A --> requestAnimationFrame 3 ---> callback B, callback C ----> requestAnimationFrame 4.

The story would be different if this was a webworker modifying shared data. But this is not the case.
So the callbacks of an event are always before/after a frame.